File: Java/Beginner/B_Greedy/BigNumber.java
package Beginner.B_Greedy;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;

import java.util.StringTokenizer;

public class BigNumber {
    private static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    private static StringTokenizer st;
    private static int[] arr;

    public static void main(String[] args) throws IOException {
        String[] tmp = br.readLine().split(" ");
        int n = Integer.parseInt(tmp[0]);
        int m = Integer.parseInt(tmp[1]);
        int k = Integer.parseInt(tmp[2]);

        /*
        st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        int k = Integer.parseInt(st.nextToken());
        */

        int idx = 0;
        arr = new int[n];

        st = new StringTokenizer(br.readLine());
        while (st.hasMoreTokens())
            arr[idx++] = Integer.parseInt(st.nextToken());

        Arrays.sort(arr);

        // 가장 큰 수가 더해지는 횟수
        int count = k * (m / (k + 1)) + m % (k + 1);         // m = 8, k = 3 -> aaabaaab

        int rst = (count * arr[n - 1]) + ((m - count) * arr[n - 2]);
        System.out.println(rst);

        br.close();
    }
}

File: v1/Python/beginner/02_Greedy/bigNumber.py
import sys

input = sys.stdin.readline 

# 큰수의 법칙
# 주어진 수들을 M번 더하여 가장 큰 수를 만드는 법칙
# 단 특정 인덱스에 해당하는 수는 연속해서 K번까지 덧셈 가능

N, M, K = map(int, input().split())
arr = list(map(int, input().split()))

arr.sort(reverse=True)

# 단순 풀이
def solution1():
    global M

    rst = 0
    cnt = 0

    while (M):
        if cnt == 3:
            rst += arr[1]
            cnt = 0
        else:
            rst += arr[0]
            cnt += 1

        M -= 1

    print(rst)


# 수열을 이용한 풀이
# {(K * arr[0]) + arr[1]}의 길이가 (K + 1)이 반복되는 형태 
# 나머지는 arr[0]을 더해주면 됨 
a = (((K * arr[0]) + arr[1]) * (M // (K + 1)))
b = arr[0] * (M % (K + 1))

rst = a + b
print(rst)